sklearn/datasets/samples_generator.py

```diff
@@ -754,35 +760,73 @@ def make_blobs(n_samples=100, n_features=2, centers=3, cluster_std=1.0,
     generator = check_random_state(random_state)
 
-    if isinstance(centers, numbers.Integral):
-        centers = generator.uniform(center_box[0], center_box[1],
-                                    size=(centers, n_features))
+    if isinstance(n_samples, numbers.Integral):
+        # n_centers is figured by the centers arg
+        if centers is None:
+            centers = 3
+
+        if isinstance(centers, numbers.Integral):
+            n_centers = centers
+            centers = generator.uniform(center_box[0], center_box[1],
+                                        size=(n_centers, n_features))
+
+        else:
+            centers = check_array(centers)
+            n_features = centers.shape[1]
+            n_centers = centers.shape[0]
+
     else:
-        centers = check_array(centers)
-        n_features = centers.shape[1]
+        # n_centers is figured by the [n_samples] arg
+        n_centers = len(n_samples)
+        if centers is None:
+            # generate centers based on [n_samples]
+            centers = generator.uniform(center_box[0], center_box[1],
+                                        size=(n_centers, n_features))
+        try:
+            assert len(centers) == n_centers
+        except TypeError:
+            raise ValueError("Parameter `centers` must be array-like.")
+        except AssertionError:
+            raise ValueError("Length of `n_samples` not consistent"
+                             " with number of centers.")
+        else:
+            centers = check_array(centers)
+            n_features = centers.shape[1]
+
+    # stds: if cluster_std is given as list, it must be consistent
+    # with the n_centers
+    if (hasattr(cluster_std, "__len__") and len(cluster_std) != n_centers):
+        raise ValueError("Length of `clusters_std` not consistent"
+                         " with number of centers.")
 
     if isinstance(cluster_std, numbers.Real):
         cluster_std = np.ones(len(centers)) * cluster_std
 
     X = []
     y = []
 
-    n_centers = centers.shape[0]
-    n_samples_per_center = [int(n_samples // n_centers)] * n_centers
+    if isinstance(n_samples, list) or isinstance(n_samples, np.ndarray):
+        n_samples_per_center = n_samples
+    else:
+        n_samples_per_center = [int(n_samples // n_centers)] * n_centers
 
-    for i in range(n_samples % n_centers):
-        n_samples_per_center[i] += 1
+        for i in range(n_samples % n_centers):
+            n_samples_per_center[i] += 1
 
     for i, (n, std) in enumerate(zip(n_samples_per_center, cluster_std)):
-        X.append(centers[i] + generator.normal(scale=std,
-                                               size=(n, n_features)))
+        X.append(generator.normal(loc=centers[i], scale=std,
+                                  size=(n, n_features)))
         y += [i] * n
 
     X = np.concatenate(X)
     y = np.array(y)
 
     if shuffle:
-        indices = np.arange(n_samples)
+        if isinstance(n_samples, list):
```

Comment:
use `np.sum` and it will work for both case

``` python
total_n_samples = np.sum(n_samples)
```

Reply:
Thanks!